app/scripts/lib/rpc-method-middleware/handlers/wallet-upgrade-account.ts

```diff
@@ -0,0 +1,289 @@
+  PendingJsonRpcResponse,
+  Json,
+} from '@metamask/utils';
+import type {
+  AsyncJsonRpcEngineNextCallback,
+  JsonRpcEngineEndCallback,
+} from '@metamask/json-rpc-engine';
+import type { SnapId } from '@metamask/snaps-sdk';
+import { toHex } from '@metamask/controller-utils';
+
+import { isAccountUpgraded } from '../../../../../shared/lib/eip7702-utils';
+import { isSnapPreinstalled } from '../../../../../shared/lib/snaps/snaps';
+// eslint-disable-next-line import/no-restricted-paths
+import { isFlask } from '../../../../../ui/helpers/utils/build-types';
+
+export type UpgradeAccountParams = {
+  account: string; // Address of the EOA to upgrade
+  chainId?: number; // Optional: chain ID for the upgrade (defaults to current)
+};
+
+export type UpgradeAccountResult = {
+  transactionHash: string; // Hash of the EIP-7702 authorization transaction
+  upgradedAccount: string; // Address of the upgraded account (same as input)
+  delegatedTo: string; // Address of the contract delegated to (determined by wallet)
+};
+
+export type GetAccountUpgradeStatusParams = {
+  account: string; // Address of the account to check
+  chainId?: number; // Optional: chain ID for the check (defaults to current)
+};
+
+export type GetAccountUpgradeStatusResult = {
+  account: string; // Address of the checked account
+  isUpgraded: boolean; // Whether the account is upgraded
+  chainId: number; // Chain ID where the check was performed
+};
+
+export const upgradeAccountHandler = {
+  methodNames: ['wallet_upgradeAccount'],
+  implementation: upgradeAccountImplementation,
+  hookNames: {
+    upgradeAccount: true,
+    getCurrentChainId: true,
+    isAtomicBatchSupported: true,
+  },
+};
+
+export const getAccountUpgradeStatusHandler = {
+  methodNames: ['wallet_getAccountUpgradeStatus'],
+  implementation: getAccountUpgradeStatusImplementation,
+  hookNames: {
+    getCurrentChainId: true,
+    getCode: true,
+    getNetworkConfigurationByChainId: true,
+  },
+};
+
+async function upgradeAccountImplementation(
+  req: JsonRpcRequest<Json[]> & { origin: string },
+  res: PendingJsonRpcResponse<Json>,
+  _next: AsyncJsonRpcEngineNextCallback,
+  end: JsonRpcEngineEndCallback,
+  {
+    upgradeAccount,
+    getCurrentChainId,
+    isAtomicBatchSupported,
+  }: {
+    upgradeAccount: (
+      address: string,
+      upgradeContractAddress: string,
+      chainId?: number,
+    ) => Promise<{ transactionHash: string; delegatedTo: string }>;
+    getCurrentChainId: () => number;
+    isAtomicBatchSupported: (request: {
+      address: string;
+      chainIds: string[];
+    }) => Promise<
+      {
+        chainId: string;
+        isSupported: boolean;
+        delegationAddress?: string;
+        upgradeContractAddress?: string;
+      }[]
+    >;
+  },
+) {
+  const { params, origin } = req;
+
+  // Check if the origin is a preinstalled snap. If its flask we allow any origin for local development.
+  if (!isFlask() && !isSnapPreinstalled(origin as SnapId)) {
```

Comment:
I like that we're being more permissive In flask, but should we disallow requests from any origins that are not valid snapIds? I'm not sure if there's value in exposing the method directly to dapps in development, because that's not possible in the dist build (snaps exports `isSnapId` method).